src/FTThresholdServer.cpp

```diff
@@ -0,0 +1,170 @@
+  mTareActionClient =
+      std::unique_ptr<TareActionClient>(new TareActionClient(nh, tareTopic));
+
+  // start action server
+  mFTThresholdActionServer.reset(
+      new actionlib::ActionServer<SetFTThresholdAction>{
+          nh, serverName,
+          std::bind(&FTThresholdServer::setForceTorqueThreshold,
+                    this, std::placeholders::_1),
+          false});
+  mFTThresholdActionServer->start();
+
+  // Set initial threshold to sensor limit
+  mForceThreshold.store(forceLimit);
+  mTorqueThreshold.store(torqueLimit);
+
+  // Do not require initial taring
+  mTaringCompleted.store(true);
+  mTimeOfLastSensorDataReceived = std::chrono::steady_clock::now();
+}
+
+//=============================================================================
+// Max F/T sensor wait time
+static const std::chrono::milliseconds MAX_DELAY =
+    std::chrono::milliseconds(100);
+
+bool FTThresholdServer::shouldStopExecution(std::string &message) {
+  // we must guarantee taring completes before moving
+  if (!mTaringCompleted.load()) {
+    ROS_WARN("taring not yet completed!");
+    return true;
+  }
+
+  // Watchdog
+  if ((std::chrono::steady_clock::now() - mTimeOfLastSensorDataReceived >
+       MAX_DELAY)) {
+    message = "Lost connection to F/T sensor!";
+    ROS_WARN(message.c_str());
+    return true;
+  }
+
+  double forceThreshold = mForceThreshold.load();
+  double torqueThreshold = mTorqueThreshold.load();
+
+  std::lock_guard<std::mutex> lock(mForceTorqueDataMutex);
+  bool forceThresholdExceeded = (mForce.norm() >= forceThreshold) && (forceThreshold != 0.0);
+  bool torqueThresholdExceeded = (mTorque.norm() >= torqueThreshold) && (torqueThreshold != 0.0);
+
+  if (forceThresholdExceeded) {
+    std::stringstream messageStream;
+    messageStream << "Force Threshold exceeded!   Threshold: " << forceThreshold
+                  << "   Force: " << mForce.x() << ", " << mForce.y() << ", "
+                  << mForce.z();
+    message = messageStream.str();
+    ROS_WARN(message.c_str());
+  }
+  if (torqueThresholdExceeded) {
+    std::stringstream messageStream;
+    messageStream << "Torque Threshold exceeded!   Threshold: "
+                  << torqueThreshold << "   Torque: " << mTorque.x() << ", "
+                  << mTorque.y() << ", " << mTorque.z();
+    message = messageStream.str();
+    ROS_WARN(message.c_str());
+  }
+
+  return forceThresholdExceeded || torqueThresholdExceeded;
+}
+
+//=============================================================================
+void FTThresholdServer::forceTorqueDataCallback(
+    const geometry_msgs::WrenchStamped &msg) {
+  std::lock_guard<std::mutex> lock(mForceTorqueDataMutex);
+  mForce.x() = msg.wrench.force.x;
+  mForce.y() = msg.wrench.force.y;
+  mForce.z() = msg.wrench.force.z;
+  mTorque.x() = msg.wrench.torque.x;
+  mTorque.y() = msg.wrench.torque.y;
+  mTorque.z() = msg.wrench.torque.z;
+  mTimeOfLastSensorDataReceived = std::chrono::steady_clock::now();
```

Comment:
The bigger issue is mismatched time bases. This is meant to be a watchdog, killing the trajectory if the FT Sensor stops broadcasting. This means that, at each loop, this value needs to compared to the current time in the local time base. If we set it to the timestamp from the Wrench topic, any mismatch between the time base of the machine that is broadcasting the topic and the current machine will eat into that Watchdog time, potentially leading to unwanted behavior. This is actually not unlikely to happen given that we aren't guaranteed to be connected to the Internet when using the robot.

Now, it may be a good idea to also add a latency check (i.e. if it takes more than like 10ms for the message to reach us from the FT sensor, then shut everything down), and that would compare the wrench header to the local time base. But I think that's a different check.